src/java/org/apache/cassandra/serializers/ListSerializer.java

```diff
@@ -255,4 +256,33 @@ public Range<Integer> getIndexesRangeFromSerialized(ByteBuffer collection,
     {
         throw new UnsupportedOperationException();
     }
+
+    @Override
+    public void forEach(ByteBuffer input, ProtocolVersion version, Consumer<ByteBuffer> action)
+    {
+        try
+        {
+            int s = readCollectionSize(input, ByteBufferAccessor.instance, ProtocolVersion.V3);
+            int offset = sizeOfCollectionSize(s, ProtocolVersion.V3);
+
+            for (int i = 0; i < s; i++)
+            {
+                int size = ByteBufferAccessor.instance.getInt(input, offset);
+                if (size < 0)
+                    continue;
+
+                offset += TypeSizes.INT_SIZE;
+
+                ByteBuffer value = ByteBufferAccessor.instance.slice(input, offset, size);
+
+                action.accept(value);
+
+                offset += size;
```

Comment:
ListSerializer inherit from CollectionSerializer so it should be possible to use the same logic that the one used for SetSerializer. Am I missing something?

Reply:
Indeed we can use the same implementation for lists and sets, moved to `CollectionSerializer` as suggested.